docs/architecture/ADR-002-universal-share-encoding.md

```diff
@@ -0,0 +1,161 @@
+# ADR 009: Universal Share Encoding
+<!-- disable markdownlint MD010 because Go code snippet uses tabs -->
+<!-- markdownlint-disable MD010 -->
+
+## Changelog
+
+- 2022/9/22: inital draft of InfoReservedByte
+- 2022/9/24: update draft to Universal Share Encoding
+
+## Context
+
+- **nid** (8 bytes): namespace id
+- **reserved** (1 byte): is the location of the first transaction, ISR, or evidence in this share if there is one and `0` if there isn't one
+- **message length** (varint 1 to 10 bytes): is the length of the entire message in bytes
+
+The current reserved namespace (transaction, ISRs, evidence) share format is:<br>`nid (8 bytes) | reserved (1 byte) | share data`
+
+The current unreserved namespace (message) share format is:
+
+- First share of message:<br>`nid (8 bytes) | message length (varint 1 to 10 bytes) | share data`
+- Contiguous share in message:<br>`nid (8 bytes) | share data`
+
+The current share format poses multiple challenges:
+
+1. Clients must have two share parsing implementations (one for reserved namespace shares and one for unreserved namespace shares).
+1. It is difficult to make changes to the share format in a backwards compatible way because clients can't determine which version of the share format an individual share conforms to.
+1. It is not possible for a client that samples a random share to determine if the share is the first share of that namespace or a contiguous share.
```

Comment:
By the way the use case isn't really about the client sampling a random share. It's more that nodes can prove that specific message in a namespace is "complete" (i.e. the node was provided all the data for that message).

Reply:
thanks for clarifying. If I understand correctly, the situation where message start indicator is of interest:
- multiple messages with the same namespace are included in the same block
- a client wants to know if one of the messages in this namespace was included in the block
- (potentially) the share of interest spans multiple shares

## illustrated example

```golang
var ns = []byte{1, 2, 3, 4, 5, 6, 7, 8}
var version = `0000000`
```
| share number            | 10      | 11      | 12              | 13      |
| ----------------------- | ------- | ------- | --------------- | ------- |
| namespace               | ns      | ns      | ns              | ns      |
| version                 | version | version | version         | version |
| message start indicator | `1`     | `1`     | `0`             | `1`     |
| data                    | foo     | bar     | bar (continued) | buzz    |


if a client is provided share 11 (with the universal share format) the client can parse the message length delim in share 11, then they know that the **bar** message will complete after reading N bytes (where N includes shares 11 and 12). This was previously impossible because a client couldn't know whether the message length delim was included in a share without parsing previous shares.

An alternative way to prove the same thing that doesn't rely on parsing the message length delim is to include a continuation share bit that is `1` when there is a continuation share after the current share and `0` when the current share ends a message. This option may be favorable in environments where parsing varints is difficult but that motivation seems disproved by https://github.com/celestiaorg/celestia-core/issues/757#issuecomment-1135592765

Reply:
Yes, exactly.

Reply:
Thanks for confirming. I added the example in https://github.com/celestiaorg/celestia-app/pull/660/commits/82191cfa0e5fe5c02e49180aaea2f5726a3955bf